astropy/units/format/utils.py

```diff
@@ -4,13 +4,27 @@
 Utilities shared by the different formats.
 """
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING
 
 from astropy.units.utils import maybe_simple_fraction
 from astropy.utils.misc import did_you_mean
 
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable, Mapping, Sequence
+    from numbers import Real
+    from typing import TypeVar
+
+    from astropy.units import UnitBase
+
+    UnitT = TypeVar("UnitT", bound=UnitBase)
+
 
-def get_grouped_by_powers(bases, powers):
+def get_grouped_by_powers(
```

Comment:
It's fine if you consider this out of scope, but perhaps then we should open an issue...
This function uses `.append`. It looks like with a simple helper function (defined externally to this function) we could use a list comprehension instead. I wonder if that would be faster — `append` versus function call overhead.

Reply:
note that list comprehensions embed a function call in Python 3.11 and older, so I would expect a gain in that area only on Python 3.12 at the moment.

Reply:
This is off topic here because type annotations don't depend on whether a function builds a `list` with repeated `.append()` calls or with a list comprehension. But it should be pointed out that using list comprehensions in this function would necessitate looping over `bases` and `powers` twice.

Reply:
On a related note. Are tuple comprehensions faster? Why doesn't this function return a tuple?

Reply:
The current code loops over `bases` and `powers` once to construct two lists. If you want to use comprehensions here you would have to loop over `bases` and `powers` twice and I don't see why that would be more performant.

Reply:
https://stackoverflow.com/a/51449836.
It can be. List comprehensions happen at the C level. `.append` happens in Python.

Reply:
> This is off topic here because type annotations don't depend on whether a function builds a list with repeated .append() calls or with a list comprehension.

"It's fine if you consider this out of scope,", so AOK by me.
Just while we had eyes on this function! Should make some kind of a decision, even if it's just to open an Issue.